packages/patrol_cli/lib/src/coverage/run_code_coverage.dart

```diff
@@ -0,0 +1,190 @@
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:coverage/coverage.dart';
+import 'package:glob/glob.dart';
+import 'package:patrol_cli/src/base/logger.dart';
+import 'package:patrol_cli/src/devices.dart';
+import 'package:vm_service/vm_service.dart';
+import 'package:vm_service/vm_service_io.dart';
+
+Future<Map<String, HitMap>> _collectCoverage(
+  VmService client,
+  Uri vmUri,
+  String packageName,
+  String mainIsolateId,
+) async {
+  final coverage = await collect(
+    vmUri,
+    false,
+    false,
+    false,
+    {packageName},
+  );
+
+  final socket = await WebSocket.connect(client.wsUri!)
+    ..add(
+      jsonEncode(
+        {
+          'jsonrpc': '2.0',
+          'id': 21,
+          'method': 'ext.patrol.markTestCompleted',
+          'params': {
+            'isolateId': mainIsolateId,
+            'command': 'markTestCompleted',
+          },
+        },
+      ),
+    );
+  await socket.close();
+
+  final map = await HitMap.parseJson(
+    coverage['coverage'] as List<Map<String, dynamic>>,
+  );
+
+  return map;
+}
+
+Future<ProcessResult> _forwardAdbPort(String host, String guest) async {
+  return Process.run('adb', ['forward', 'tcp:$host', 'tcp:$guest']);
+}
+
+Uri _createWebSocketUri(Uri uri) {
+  final pathSegments = uri.pathSegments.where((c) => c.isNotEmpty).toList()
+    ..add('ws');
+  return uri.replace(scheme: 'ws', pathSegments: pathSegments);
+}
+
+Future<void> _saveCoverage(String report) async {
+  final coverageDirectory = Directory('coverage');
+
+  if (!coverageDirectory.existsSync()) {
+    await coverageDirectory.create();
+  }
+  await File(
+    coverageDirectory.uri.resolve('patrol_lcov.info').toString(),
+  ).writeAsString(report);
+}
+
+Future<void> runCodeCoverage({
+  required String flutterPackageName,
+  required Directory flutterPackageDirectory,
+  required TargetPlatform platform,
+  required Logger logger,
+  required Set<Glob> ignoreGlobs,
+}) async {
+  final homeDirectory =
+      Platform.environment['HOME'] ?? Platform.environment['USERPROFILE'];
+
+  final logsProcess = await Process.start(
+    'flutter',
+    ['logs'],
+    workingDirectory: homeDirectory,
+  );
+  final vmRegex = RegExp('listening on (http.+)');
+
+  final hitMap = <String, HitMap>{};
+  int? totalTestCount;
+  var count = 0;
+
+  logsProcess.stdout.transform(utf8.decoder).listen(
+    (line) async {
+      final vmLink = vmRegex.firstMatch(line)?.group(1);
+
+      if (vmLink == null) {
+        return;
+      }
+
+      final port = RegExp(':([0-9]+)/').firstMatch(vmLink)!.group(1)!;
+      final auth = RegExp(':$port/(.+)').firstMatch(vmLink)!.group(1);
+
+      final String? hostPort;
+
+      switch (platform) {
+        case TargetPlatform.android:
+          await _forwardAdbPort('61011', port);
```

Comment:
where does this number come from?

Reply:
fully random, should we just pick a port and use it or search for an unused one at runtime?

Reply:
That's okay, let's just extract it to a constant and add a doc comments saying it's a randomly selected number.

Reply:
Actually, I think we should search for unused port instead of hardcoding it. We already have some issues with other ports being hardcoded so we should avoid it whenever possible

Reply:
done, please have a look